File: 0930-binary-subarrays-with-sum/0930-binary-subarrays-with-sum.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        
        def atleastk(k):
            left = 0
            res = 0
            count = 0
            for right in range(len(nums)):
                res+=nums[right]
                
                while left <= right and res > k:
                    res-=nums[left]
                    left+=1
                count+=(right-left+1)
            return count
        ans = abs(atleastk(goal) - atleastk(goal-1))
        return ans
                
                
            

File: 0930-binary-subarrays-with-sum/0930-binary-subarrays-with-sum.cpp
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        
        
        
       
        function<int(int k)> al = [&](int k){
          int left = 0;
          long long sm = 0;
          int cnt = 0;
          for(int right = 0; right < nums.size(); right++){
            sm += nums[right];
              
            while(left <= right && sm > k){
                sm-=nums[left];
                ++left;
            }
            
            cnt+=(right - left + 1);
          }  
        return cnt;
        };
        
        
        return abs(al(goal) - al(goal -1));
    }
};